### ISSUE TITLE
Support B2C multi scopes access for different APIs

### ISSUE THREAD

**Issue by paszkowskik** [damienbod/angular-auth-oidc-client#142]
Hi, we have a problem that our angular client need to have an access for two different APIs based on azure B2C (both have different scopes). B2C has policy that each resource (scope) must have its own token so we can't use one access token to authorize two different api.
Our suggestion is support B2C multi scopes that will generate token for each scope. We are concious that its a lot of effort from your side so we propose to prepare that solution and create pull request during next week.
What do you think about that idea ?

**Comment by damienbod** [comment]
@paszkowskik 

Sounds good. I think this could be implemented pretty easy. The first login requests the id_token token (one scope). The the app could then request a second token using the token endpoint.

opinion?

Greetings Damien

**Comment by paszkowskik** [comment]
I am not sure that I understand your vision, can you explain it more ?

**Comment by damienbod** [comment]
Hopefully I understood correctly :)

You need 2 access tokens, on per API. Thhe first access token can be got by doing the id_token token flow. Only request 1 scope. Then after a successfull login, you can use this access_token and request a second token using the OIDC token endpoint, wich returns the second token with the other scope.

The package would then need to support the OIDC token endpoint and then you would need to add application logic. This would be a extra HTTP request after a successful login. And you continue to use standard OIDC flows

? Just a suggestion.

What were you thinking?
Greetings Damien

**Comment by paszkowskik** [comment]
Now we understand it clear. Solution that you propose is simple and exacly that we need. Only one problem more is how to handle scopes config (maybe it should has names like key - value) and corelation of conqret scope with conqret access token. 
Can you provide features that you described in previous post ?

**Comment by damienbod** [comment]
Yes, but it might take a while, I'm very busy at my day job, will try to get it done before XMAS, unless you do a PR first

Thanks for the feedback.

greetings Damien

**Comment by damienbod** [comment]
@paszkowskik Just 1 thing, you should check if the B2C STS supports this first. Is this possible to config the token endpoint in this way? I will implement this using IdentityServer4 as the STS. If we use the OIDC token endpoint, it should work without any further changes.

**Comment by paszkowskik** [comment]
I am reading now how we can implement that feature and I have found usefull article. There is written that we need to use authorization code flow than after code gain you have redeem that on access token.
https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-reference-oauth-code

GET https://login.microsoftonline.com/fabrikamb2c.onmicrosoft.com/oauth2/v2.0/authorize?
client_id=90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6
&response_type=code
&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob
&response_mode=query
&scope=90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6%20offline_access
&state=arbitrary_data_you_can_receive_in_the_response
&p=b2c_1_sign_in

"Now that you've acquired an authorization code, you can redeem the code for a token to the intended resource by sending a POST request to the /token endpoint. In Azure AD B2C, the only resource that you can request a token for is your app's own back-end web API. The convention that's used for requesting a token to yourself is to use your app's client ID as the scope"

POST fabrikamb2c.onmicrosoft.com/oauth2/v2.0/token?p=b2c_1_sign_in HTTP/1.1
Host: https://login.microsoftonline.com
Content-Type: application/x-www-form-urlencoded

grant_type=authorization_code&client_id=90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6&scope=90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6 offline_access&code=AwABAAAAvPM1KaPlrEqdFSBzjqfTGBCmLdgfSTLEMPGYuNHSUYBrq...&redirect_uri=urn:ietf:wg:oauth:2.0:oob

**Comment by paszkowskik** [comment]
Yes we will check it today to be sure about possible implementation by /token endpoint

**Comment by paszkowskik** [comment]
Its seems like B2C its not the same like  Identity Server 4 because propably we cannot use /token endpoint to get access token for another scope. We are still analyse that how it works in B2C. Currently we have to hold that work because from tommorow everyone has holidays until new year.

**Comment by paszkowskik** [comment]
Could you tell us which grant type you would to use in that case ?

**Comment by damienbod** [comment]
@paszkowskik Sorry missed this issue... 

Do you still have questions here?

Greetings Damien

**Comment by paszkowskik** [comment]
Hi, sorry for late answer. We have changed the library few months ago. We experienced problem with token refresh, especially on IE it hang up whole app during token refresh for 3 seconds. it was because whole app was loaded into iframe (it shouldnt work like that). We changed library where silent refresh is implemented correctly.
